libvast/src/address.cpp

```diff
@@ -16,13 +16,104 @@
 
 #include <arpa/inet.h>
 #include <netinet/in.h>
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/ossl_typ.h>
 #include <sys/socket.h>
 
 #include <cstdlib>
 #include <utility>
 
 namespace vast {
 
+namespace {
+
+inline uint32_t bitmask32(size_t bottom_bits) {
+  return bottom_bits >= 32 ? 0xffffffff : ((uint32_t{1} << bottom_bits) - 1);
+}
+
+class address_encryptor {
+public:
+  address_encryptor(const std::array<address::byte_type, 32>& key) {
+    ctx_ = EVP_CIPHER_CTX_new();
+    EVP_CIPHER_CTX_init(ctx_);
+    OpenSSL_add_all_ciphers();
+    cipher_ = EVP_get_cipherbyname("aes-128-ecb");
+    block_size_ = EVP_CIPHER_block_size(cipher_);
+    pad_ = std::vector<address::byte_type>(block_size_);
+    auto pad_out_len = 0;
+    EVP_CipherInit_ex(ctx_, cipher_, nullptr, key.data(), nullptr, 1);
+    // use second 16-byte half of key for padding
+    EVP_CipherUpdate(ctx_, pad_.data(), &pad_out_len, key.data() + block_size_,
+                     block_size_);
+  }
+
+  address_encryptor(const address_encryptor&) = delete;
+
+  ~address_encryptor() {
+    EVP_CIPHER_CTX_cleanup(ctx_);
+    EVP_CIPHER_CTX_free(ctx_);
+  }
```

Comment:
There are two design bugs in the highlighted code:
- Destructors, move constructors, and move-assignment operators should always be marked noexcept (see [here](https://quuxplusone.github.io/blog/2022/08/26/vector-pessimization/))
- Per the [rule of five](https://en.cppreference.com/w/cpp/language/rule_of_three), if you define any of the five then you must also define the remaining.

You can avoid all of this by applying the rule of zero, which is much cleaner design anyways: Make `ctx_` a `std::unique_ptr` with a custom deleter that calls the two functions, and then simply remove the explicitly defaulted copy constructor and the custom destructor.

Reply:
Good idea with the `unique_ptr`, it certainly avoids me having to `delete` the rest of the constructors.
Implemented in https://github.com/tenzir/vast/pull/2719/commits/01eaddb2b09fbaeaf0cc8ead313af0af0ef9bd46